PowerPointLabs/PowerPointLabs/FunctionalTestInterface.Impl/Controller/ColorsLabController.cs

```diff
@@ -16,114 +23,364 @@ class ColorsLabController : MarshalByRefObject, IColorsLabController
+                {
+                    rect.RaiseEvent(
+                        new MouseButtonEventArgs(Mouse.PrimaryDevice, Environment.TickCount, MouseButton.Left)
+                        {
+                            RoutedEvent = UIElement.MouseDownEvent
+                        });
+
+                    rect.RaiseEvent(
+                       new MouseButtonEventArgs(Mouse.PrimaryDevice, Environment.TickCount, MouseButton.Left)
+                       {
+                           RoutedEvent = UIElement.MouseUpEvent
+                       });
+                }));
+            }
+        }
+
+        public void ClickComplementaryRect(int index)
+        {
+            System.Windows.Shapes.Rectangle rect;
+
+            switch (index)
+            {
+                case 1:
+                rect = _pane.ColorsLabPaneWPF1.complementaryLowerRectangle;
+                break;
+                case 2:
+                rect = _pane.ColorsLabPaneWPF1.complementaryMiddleRectangle;
+                break;
+                case 3:
+                rect = _pane.ColorsLabPaneWPF1.complementaryHigherRectangle;
+                break;
+                default:
+                rect = null;
+                break;
+            }
+
+            if (_pane != null && rect != null)
+            {
+                UIThreadExecutor.Execute((Action)(() =>
+                {
+                    rect.RaiseEvent(
+                        new MouseButtonEventArgs(Mouse.PrimaryDevice, Environment.TickCount, MouseButton.Left)
+                        {
+                            RoutedEvent = UIElement.MouseDownEvent
+                        });
+
+                    rect.RaiseEvent(
+                       new MouseButtonEventArgs(Mouse.PrimaryDevice, Environment.TickCount, MouseButton.Left)
+                       {
+                           RoutedEvent = UIElement.MouseUpEvent
+                       });
+                }));
+            }
+        }
+
+        public void ClickTriadicRect(int index)
+        {
+            System.Windows.Shapes.Rectangle rect;
+
+            switch (index)
+            {
+                case 1:
+                rect = _pane.ColorsLabPaneWPF1.triadicLowerRectangle;
+                break;
+                case 2:
+                rect = _pane.ColorsLabPaneWPF1.triadicMiddleRectangle;
+                break;
+                case 3:
+                rect = _pane.ColorsLabPaneWPF1.triadicHigherRectangle;
+                break;
+                default:
+                rect = null;
+                break;
+            }
+
+            if (_pane != null && rect != null)
+            {
+                UIThreadExecutor.Execute((Action)(() =>
+                {
+                    rect.RaiseEvent(
+                        new MouseButtonEventArgs(Mouse.PrimaryDevice, Environment.TickCount, MouseButton.Left)
+                        {
+                            RoutedEvent = UIElement.MouseDownEvent

```

Comment:
This routed event code seems to be repeated quite a few times, would it be possible to refactor into a function?

Reply:
Good idea, done!